File: pyfenstein3d/engine/block.py
import math
from .vector2d import Vector2d
from abc import abstractmethod
from abc import abstractclassmethod

class Block():
    def __init__(self, vector_x: float, vector_y: float, type_id: int, is_solid: bool=False, is_moveable: bool=False):
        self._vector2d = Vector2d(vector_x, vector_y)
        self.__type_id = type_id
        self._is_solid = is_solid
        self.__is_moveable = is_moveable
        self._set_x_y(vector_x, vector_y)

    @property
    def x(self):
        return self._vector2d.x

    @property
    def y(self):
        return self._vector2d.y

    @property
    def block_x(self):
        return self.__block_x

    @property
    def block_y(self):
        return self.__block_y

    @property
    def offset_x(self):
        return self.__offset_x

    @property
    def offset_y(self):
        return self.__offset_y

    @property
    def type_id(self):
        return self.__type_id

    @property
    def is_solid(self):
        return self._is_solid

    @property
    def is_moveable(self):
        return self.__is_moveable

    def _set_x_y(self, vector_x: float, vector_y: float):
        self._vector2d = Vector2d(vector_x, vector_y)
        self.__block_x = math.floor(vector_x)
        self.__block_y = math.floor(vector_y)
        self.__offset_x = vector_x % 1
        self.__offset_y = vector_y % 1

    @abstractmethod
    def _get_bounds(self, fov_ang: float):
        pass

    def is_in_fov(self, pos: Vector2d, fov_ang: float, fov_delta: float, dist: float):
        bounds = self._get_bounds(fov_ang)
        # fov_ang_min = (fov_ang - (fov_delta * 0.5) + math.pi * 2) % (math.pi * 2)
        # fov_ang_max = (fov_ang + (fov_delta * 0.5) + math.pi * 2) % (math.pi * 2)
        # fov_ang_min, fov_ang_max = (fov_ang_min, fov_ang_max) if fov_ang_min < fov_ang_max else (fov_ang_max, fov_ang_min)
        for bound in bounds:
            bound_pos = (bound - pos) ** -fov_ang
            if dist > bound_pos.mag:
                bound_ang = bound_pos.ang - math.pi * 2 if bound_pos.ang > math.pi else bound_pos.ang
                if fov_delta * -0.5 <= bound_ang <= fov_delta * 0.5:
                    return True
        return False

    
    def interacted(self):
        pass

File: go/engine/block.go
package engine

import "math"

// Blocker is the common interface for all block types in the engine.
type Blocker interface {
	X() float64
	Y() float64
	BlockX() int
	BlockY() int
	OffsetX() float64
	OffsetY() float64
	TypeID() int
	IsSolid() bool
	IsMoveable() bool
	GetBounds(fovAng float64) []*Vector2d
	IsInFov(pos *Vector2d, fovAng, fovDelta, dist float64) bool
	Interacted()
	GetState(player *Player) int
}

// ItemLike is implemented by Item and all its subclasses (has a Touch method).
type ItemLike interface {
	Blocker
	Touch(player *Player, grid *ItemGrid)
}

// block holds common position and type data for all block types.
type block struct {
	pos        *Vector2d
	typeID     int
	isSolid    bool
	isMoveable bool
	blockX     int
	blockY     int
	offsetX    float64
	offsetY    float64
}

func newBlock(x, y float64, typeID int, isSolid, isMoveable bool) block {
	b := block{
		pos:        NewVector2d(x, y),
		typeID:     typeID,
		isSolid:    isSolid,
		isMoveable: isMoveable,
	}
	b.setXY(x, y)
	return b
}

func (b *block) setXY(x, y float64) {
	b.pos = NewVector2d(x, y)
	b.blockX = int(math.Floor(x))
	b.blockY = int(math.Floor(y))
	b.offsetX = math.Mod(x, 1)
	b.offsetY = math.Mod(y, 1)
}

func (b *block) X() float64      { return b.pos.X() }
func (b *block) Y() float64      { return b.pos.Y() }
func (b *block) BlockX() int     { return b.blockX }
func (b *block) BlockY() int     { return b.blockY }
func (b *block) OffsetX() float64 { return b.offsetX }
func (b *block) OffsetY() float64 { return b.offsetY }
func (b *block) TypeID() int     { return b.typeID }
func (b *block) IsSolid() bool   { return b.isSolid }
func (b *block) IsMoveable() bool { return b.isMoveable }
func (b *block) Interacted()     {}
func (b *block) GetState(player *Player) int { return 0 }

// blockIsInFov is a helper used by all concrete types that implement IsInFov.
func blockIsInFov(bounds []*Vector2d, pos *Vector2d, fovAng, fovDelta, dist float64) bool {
	for _, bound := range bounds {
		boundPos := bound.Sub(pos).Rot(-fovAng)
		if dist > boundPos.Mag() {
			boundAng := boundPos.Ang()
			if boundAng > math.Pi {
				boundAng -= math.Pi * 2
			}
			if -fovDelta*0.5 <= boundAng && boundAng <= fovDelta*0.5 {
				return true
			}
		}
	}
	return false
}
